### PR TITLE
kube: report errors back to user on every failure

### PR DESCRIPTION
**Pull request by tigrato**
This PR fixes an issue where the client could miss the exit error because the session was abruptly closed.

This PR adds coverage to ensure recorder failures force the session to be closed.

### CODE REVIEW

File: lib/kube/proxy/sess.go [hunk 1]
```diff
@@ -547,14 +552,22 @@ func (s *session) launch(isEphemeralCont bool) (returnErr error) {
...
 	defer func() {
+		onErr := func(err error) {
+			s.setTerminationErr(err)
+			s.reportErrorToSessionRecorder(err)
+			s.log.WithError(err).Warning("Executor failed while streaming.")
+		}
+
+		onErr(returnErr)
 		// call onFinished to emit the session.end and exec events.
+		// onFinish is never nil
```
Comment (codingllama) [inline lib/kube/proxy/sess.go:565]: nit:

```suggestion
		// onFinish is never nil.
```

File: lib/kube/proxy/sess.go [hunk 2]
```diff
@@ -547,14 +552,22 @@ func (s *session) launch(isEphemeralCont bool) (returnErr error) {
...
 	eventPodMeta := request.eventPodMeta(request.context, s.sess.kubeAPICreds)
 
-	onFinished, err := s.lockedSetupLaunch(request, q, eventPodMeta)
-	if err != nil {
-		return trace.Wrap(err)
-	}
+	onFinished, err := s.lockedSetupLaunch(request, eventPodMeta)
 	defer func() {
+		onErr := func(err error) {
+			s.setTerminationErr(err)
```
Comment (codingllama) [inline lib/kube/proxy/sess.go:558]: Should we check `err != nil` before running the code below, particularly the log statement?

File: lib/kube/proxy/sess.go [hunk 3]
```diff
@@ -547,14 +552,22 @@ func (s *session) launch(isEphemeralCont bool) (returnErr error) {
...
+			s.reportErrorToSessionRecorder(err)
+			s.log.WithError(err).Warning("Executor failed while streaming.")
+		}
+
+		onErr(returnErr)
```
Comment (codingllama) [inline lib/kube/proxy/sess.go:563]: Suggestion: ditch the additional func and move the code here?

File: lib/kube/proxy/sess.go [hunk 4]
```diff
@@ -921,7 +863,57 @@ func (s *session) lockedSetupLaunch(request *remoteCommandRequest, q url.Values,
...
+							s.log.WithError(err).Error("Failed to close session")
+						}
+					}
+				case <-s.closeC:
+					return
+				}
+			}
+		}()
+	}
+	return onFinish, nil
```
Comment (codingllama) [inline lib/kube/proxy/sess.go:916]: Is this block a move or are there changes here?

File: lib/kube/proxy/sess.go [hunk 5]
```diff
@@ -921,7 +863,57 @@ func (s *session) lockedSetupLaunch(request *remoteCommandRequest, q url.Values,
...
+			ticker := time.NewTicker(PresenceVerifyInterval)
+			defer ticker.Stop()
+
+			for {
+				select {
+				case <-ticker.C:
+					err := s.checkPresence()
+					if err != nil {
+						s.log.WithError(err).Error("Failed to check presence, closing session as a security measure")
+						err := s.Close()
```
Comment (codingllama) [inline lib/kube/proxy/sess.go:905]: Should we exit the loop after the Close, or does this keep working somehow?

If it does keep working, please add a brief comment explaining it.

File: lib/kube/proxy/sess.go [hunk 6]
```diff
@@ -921,7 +863,57 @@ func (s *session) lockedSetupLaunch(request *remoteCommandRequest, q url.Values,
...
+						if err != nil {
```
Comment (codingllama) [inline lib/kube/proxy/sess.go:906]: nit: Use the smaller scope, give it a space.

```suggestion

						if err := s.Close(); err != nil {
```

File: lib/kube/proxy/sess_test.go [hunk 7]
```diff
@@ -73,99 +72,140 @@ func TestSessionEndError(t *testing.T) {
...
-	)
+		{
+			name:        "non-interactive without recording error",
+			interactive: false,
+		},
+		{
+			name:         "interactive with recording error",
+			interactive:  true,
+			recordingErr: recordingErr,
+		},
```
Comment (codingllama) [inline lib/kube/proxy/sess_test.go:92]: Should we test non-interactive with recording error?

File: lib/kube/proxy/sess_test.go [hunk 8]
```diff
@@ -73,99 +72,140 @@ func TestSessionEndError(t *testing.T) {
...
+							assert.Empty(t, execEvent.ExitCode)
+							assert.Equal(t, tt.recordingErr.Error(), execEvent.Error)
+						}
+						hasSessionExecEvent = true
+					}
+					assert.Truef(t, hasSessionEndEvent, "session end event not found in audit log")
+					assert.Truef(t, hasSessionExecEvent, "session exec event not found in audit log")
+				}, 10*time.Second, 1*time.Second)
+			} else {
+				require.Never(t, func() bool {
```
Comment (codingllama) [inline lib/kube/proxy/sess_test.go:201]: Do we need the periodic checking or is this equivalent to sleep(1s) -> check?

Could we do this in a way that doesn't take 1s?

File: lib/kube/proxy/sess.go (hunk 4 above)
Comment (tigrato) [reply lib/kube/proxy/sess.go:916]: Just a simple move.

File: lib/kube/proxy/sess.go
Comment (codingllama) [inline lib/kube/proxy/sess.go:1]: What is the change here that fixes the PR title? I'm not super familiar with lib/kube, so it's not super clear to me.

File: lib/kube/proxy/sess.go (hunk 4 above)
Comment (codingllama) [reply lib/kube/proxy/sess.go:916]: I thought so as I was eyeballing it, but thanks for confirming.

File: lib/kube/proxy/sess_test.go (hunk 7 above)
Comment (tigrato) [reply lib/kube/proxy/sess_test.go:92]: non interactive sessions aren't recorded

File: lib/kube/proxy/sess.go (hunk 5 above)
Comment (tigrato) [reply lib/kube/proxy/sess.go:905]: Nice catch. I just moved this code and didn't saw it

File: lib/kube/proxy/sess_test.go (hunk 8 above)
Comment (tigrato) [reply lib/kube/proxy/sess_test.go:201]: unfortunately no. They are async and we need to wait for collection. The test takes 5s without this

File: lib/kube/proxy/sess.go
Comment (tigrato) [reply lib/kube/proxy/sess.go:1]: This section https://github.com/gravitational/teleport/pull/42154/files#diff-a98dd59a36206bb300939248013dd2e1b713077750f5d7b7c0c7e8b2d7fa443bR555-R564

Comment (tigrato) [reply lib/kube/proxy/sess.go:1]: If you want, I can split everything but takes a bit of time

Comment (codingllama) [reply lib/kube/proxy/sess.go:1]: No need to split anything, I just wanted to clarify.

The lockedSetupLaunch() error wasn't handled by an `onErr`, right?

**Review by codingllama — APPROVED** [review]

File: lib/kube/proxy/sess.go [hunk 9]
```diff
@@ -547,14 +551,20 @@ func (s *session) launch(isEphemeralCont bool) (returnErr error) {
...
-	}
+	onFinished, err := s.lockedSetupLaunch(request, eventPodMeta)
 	defer func() {
+		if returnErr != nil {
+			s.setTerminationErr(returnErr)
+			s.reportErrorToSessionRecorder(returnErr)
+			s.log.WithError(returnErr).Warning("Executor failed while streaming.")
+		}
 		// call onFinished to emit the session.end and exec events.
+		// onFinish is never nil.
```
Comment (AntonAM) [inline lib/kube/proxy/sess.go:562]:
```suggestion
		// onFinished is never nil.
```

**Review by AntonAM — APPROVED** [review]

**Pull request merged**